### ISSUE TITLE
systemctl poweroff/reboot --force isn't working correctly: two issues

### ISSUE THREAD

**Issue by sourcejedi** [systemd/systemd#8909]
### Submission type

 - Bug report

### systemd version the issue has been seen with

systemd-238-7.fc28.1.x86_64

### Used distribution

Fedora 28

### Bug 1: not unmounting cleanly

On Fedora, PackageKit offline updates have a bug where it uses a forced shutdown, instead of a graceful shutdown. On both Fedora 27, and Fedora 28, this has caused a fsck of the root filesystem (only) on subsequent boot. I don't think this should happen, even with the forced shutdown.

https://bugzilla.redhat.com/show_bug.cgi?id=1564462

Some more detailed evidence of unclean unmount on Fedora 28 is available at https://unix.stackexchange.com/questions/441511/does-recovering-journal-prove-an-unclean-shutdown-unmount

System is pretty straightforward Fedora. A single spinning hard drive. Root filesystem on LVM.  I have no other block devices in `lsblk`. I am not using any network filesystems. I was not using any MTP mounted device or anything else which would have required FUSE.

I happen to have a self-bind mount on `/mnt`. EDIT3: the unclean unmounts definitely pre-date this bind mount.

EDIT: My historical fsck logs are annoying to look at, because systemd v234 in Fedora 27 was also not cleanly unmounting the root filesystem anyway. https://bugzilla.redhat.com/show_bug.cgi?id=1554943 EDIT2: but they seem to include a time before systemd v234, where unclean unmounts only happen during PK offline updates.  The unclean unmount (of rootfs only) seems to happen consistently in this case.

## Bug 2: Hang

I experimented to try and reproduce the above. I copied a source tree to the root filesystem (without `sync`), and immediately ran `systemctl poweroff --force` (or reboot).

Several times (but far from all) when I tried this, I observed a hang. Obviously gnome disappeared. I haven't been able to see any messages from systemd-shutdown, despite the last time using `dmesg -E` and `dmesg -n 8` just in case. The last time, the hang was at a completely black screen.  Other times, it showed the systemd messages from boot time.

I know there was no disk activity during the hang, observing both the disk activity LED and also I didn't hear any hard disk noises :).

On the last occasion, I timed the hang, it lasted about 70 seconds before rebooting. (On that occasion I had used `reboot`).

On the other occasions, I had used `poweroff`, but at least once I am certain the hang was finite (and did not reboot, only poweroff). Therefore I think we can rule out that the hang ended due to a hardware watchdog, because the hardware watchdog is only set for reboots.

---

Analysis note. Although a graceful shutdown of a Fedora LVM root system will involve a shutdown initramfs, this does not apply to a forced shutdown. The reason is that if you look at `dracut-shutdown.service`, /run/initramfs is only populated during the graceful shutdown of services. (I also noticed that on a VM with rootfs on a partition instead of LVM, dracut does not create `/run/initramfs/.need_shutdown`, and then `dracut-shutdown.service` doesn't populate `/run/initramfs` for shutdown).

**Comment by sourcejedi** [comment]
Thinking about where 70 seconds might come from...

I notice that systemd-shutdown does sync both at the start and end. And each time it waits for at least 3 intervals of 10 seconds, if it thinks sync() is still running.  So that's worth thinking about.

The timeout for terminating (and then killing) processes is 90 seconds (in v238/v237, this is taken from DEFAULT_TIMEOUT_USEC, but the value is unchanged).

CC @kyle-walker. Fedora 27 with systemd v234 was definitely affected by the unclean unmount bug.  However I only tried to reproduce the hang on Fedora 28 with systemd v238. So far I think it's reasonable to guess the finite hang occurs in one of the timeouts added in #6598 ... of course that might just mean v234 would hang forever in the same case, I don't know if the hang is a new bug.

**Comment by sourcejedi** [comment]
The last time I reproduced the 70 second shutdown hang, obviously I had let the hang finish, and the subsequent boot does *not* show "systemd-fsck[nnn]: /dev/mmm: recovering journal".

I have one more "recovering journal" on the day I first tested this though. Maybe that was when I didn't realize the hang was finite, and shut the hardware down by holding the power button.  I thought I'd done that more than once though, so it's a bit confusing :). In this instance, the message was shown for each of my 3 ext4 partitions.
